File: objectModel/Python/cdm/persistence/cdmfolder/projections/operation_base_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import TYPE_CHECKING

from cdm.enums import CdmObjectType, CdmLogCode
from cdm.enums.cdm_operation_type import OperationTypeConvertor, CdmOperationType
from cdm.utilities.logging import logger
from cdm.utilities.string_utils import StringUtils

if TYPE_CHECKING:
    from cdm.objectmodel import CdmCorpusContext, CdmOperationBase
    from cdm.utilities import ResolveOptions, CopyOptions


def _make_data_object(object_type: 'CdmObjectType'):
    """Instantiates a data object based on the object type."""
    from cdm.persistence.cdmfolder.types import OperationAddAttributeGroup, OperationAddCountAttribute, OperationAddSupportingAttribute, \
        OperationAddTypeAttribute, OperationArrayExpansion, OperationCombineAttributes, OperationExcludeAttributes, OperationIncludeAttributes, \
        OperationRenameAttributes, OperationReplaceAsForeignKey
    data_map = {
        CdmObjectType.OPERATION_ADD_ATTRIBUTE_GROUP_DEF: lambda: OperationAddAttributeGroup(),
        CdmObjectType.OPERATION_ADD_COUNT_ATTRIBUTE_DEF: lambda: OperationAddCountAttribute(),
        CdmObjectType.OPERATION_ADD_SUPPORTING_ATTRIBUTE_DEF: lambda: OperationAddSupportingAttribute(),
        CdmObjectType.OPERATION_ADD_TYPE_ATTRIBUTE_DEF: lambda: OperationAddTypeAttribute(),
        CdmObjectType.OPERATION_ARRAY_EXPANSION_DEF: lambda: OperationArrayExpansion(),
        CdmObjectType.OPERATION_COMBINE_ATTRIBUTES_DEF: lambda: OperationCombineAttributes(),
        CdmObjectType.OPERATION_EXCLUDE_ATTRIBUTES_DEF: lambda: OperationExcludeAttributes(),
        CdmObjectType.OPERATION_INCLUDE_ATTRIBUTES_DEF: lambda: OperationIncludeAttributes(),
        CdmObjectType.OPERATION_RENAME_ATTRIBUTES_DEF: lambda: OperationRenameAttributes(),
        CdmObjectType.OPERATION_REPLACE_AS_FOREIGN_KEY_DEF: lambda: OperationReplaceAsForeignKey(),
    }

    return data_map.get(object_type)()


class OperationBasePersistence:
    @staticmethod
    def from_data(ctx: 'CdmCorpusContext', object_type: 'CdmObjectType', data: 'OperationBase'):
        if data is None:
            return None

        operation = ctx.corpus.make_object(object_type)  # type: CdmOperationBase
        operation_type = OperationTypeConvertor._from_object_type(object_type)  # type: CdmOperationType
        operation_name = OperationTypeConvertor._operation_type_to_string(CdmOperationType.COMBINE_ATTRIBUTES)
        if data.type and not StringUtils.equals_with_ignore_case(data.type, operation_name):
            logger.error(ctx, operation_name, OperationBasePersistence.from_data.__name__, None, CdmLogCode.ERR_PERSIST_PROJ_INVALID_TYPE, data.type)
        else:
            operation.type = operation_type

        operation.condition = data.condition
        operation.explanation = data.explanation
        operation.source_input = data.sourceInput

        return operation

    @staticmethod
    def to_data(instance: 'CdmOperationBase', res_opt: 'ResolveOptions', options: 'CopyOptions'):
        data = _make_data_object(instance.object_type)
        data.type = OperationTypeConvertor._operation_type_to_string(instance.type)
        data.condition = instance.condition
        data.explanation = instance.explanation
        data.sourceInput = instance.source_input

        return data

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/persistence/cdmfolder/projections/OperationBasePersistence.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.projections;

import com.fasterxml.jackson.databind.JsonNode;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusContext;
import com.microsoft.commondatamodel.objectmodel.cdm.projections.CdmOperationBase;
import com.microsoft.commondatamodel.objectmodel.cdm.projections.OperationTypeConvertor;
import com.microsoft.commondatamodel.objectmodel.enums.CdmLogCode;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.enums.CdmOperationType;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationAddAttributeGroup;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationAddCountAttribute;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationAddSupportingAttribute;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationAddTypeAttribute;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationArrayExpansion;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationBase;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationCombineAttributes;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationExcludeAttributes;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationIncludeAttributes;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationRenameAttributes;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.types.projections.OperationReplaceAsForeignKey;
import com.microsoft.commondatamodel.objectmodel.utilities.CopyOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.ResolveOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.StringUtils;
import com.microsoft.commondatamodel.objectmodel.utilities.logger.Logger;

public class OperationBasePersistence {
  private static final String TAG = OperationBasePersistence.class.getSimpleName();

  public static <T extends CdmOperationBase> T fromData(final CdmCorpusContext ctx, final CdmObjectType objectType, final JsonNode obj) {
    if (obj == null) {
      return null;
    }

    CdmOperationBase operation = ctx.getCorpus().makeObject(objectType);
    CdmOperationType operationType = OperationTypeConvertor.fromObjectType(objectType);
    String operationName = OperationTypeConvertor.operationTypeToString(operationType);

    if (obj.get("$type") != null && !StringUtils.equalsWithIgnoreCase(obj.get("$type").asText(), operationName)) {
      Logger.error(ctx, TAG, "fromData", operation.getAtCorpusPath(), CdmLogCode.ErrPersistProjInvalidType, obj.get("$type").toString());
    } else {
        operation.setType(operationType);
    }

    if (obj.get("condition") != null) {
      operation.setCondition(obj.get("condition").asText());
    }

    if (obj.get("explanation") != null) {
        operation.setExplanation(obj.get("explanation").asText());
    }

    if (obj.get("sourceInput") != null) {
      operation.setSourceInput(obj.get("sourceInput").asBoolean());
    }

    return (T) operation;
  }

  public static <T extends OperationBase> T toData(final CdmOperationBase instance, ResolveOptions resOpt, CopyOptions options) {
    if (instance == null) {
      return null;
    }

    T obj = makeDataObject(instance.getObjectType());
    obj.setType(OperationTypeConvertor.operationTypeToString(instance.getType()));
    obj.setCondition(instance.getCondition());
    obj.setExplanation(instance.getExplanation());
    obj.setSourceInput(instance.getSourceInput());

    return (T)obj;
  }

  /**
   * Instantiates a data object based on the object type.
   */
  private static <T extends OperationBase> T makeDataObject(CdmObjectType objectType) {
    OperationBase obj;
    switch(objectType) {
      case OperationAddAttributeGroupDef:
        obj = new OperationAddAttributeGroup();
        break;
      case OperationAddCountAttributeDef:
        obj = new OperationAddCountAttribute();
        break;
      case OperationAddSupportingAttributeDef:
        obj = new OperationAddSupportingAttribute();
        break;
      case OperationAddTypeAttributeDef:
        obj = new OperationAddTypeAttribute();
        break;
      case OperationArrayExpansionDef:
        obj = new OperationArrayExpansion();
        break;
      case OperationCombineAttributesDef:
        obj = new OperationCombineAttributes();
        break;
      case OperationExcludeAttributesDef:
        obj = new OperationExcludeAttributes();
        break;
      case OperationIncludeAttributesDef:
        obj = new OperationIncludeAttributes();
        break;
      case OperationRenameAttributesDef:
        obj = new OperationRenameAttributes();
        break;
      case OperationReplaceAsForeignKeyDef:
        obj = new OperationReplaceAsForeignKey();
        break;
      default:
        return null;
    }

    return (T) obj;
  }
}